dulwich/sparse_patterns.py

```diff
@@ -0,0 +1,462 @@
+      force: Whether to forcibly remove locally modified files (default False).
+
+    Returns:
+      None
+    """
+    index = repo.open_index()
+    normalizer = repo.get_blob_normalizer()
+
+    def local_modifications_exist(full_path, index_entry):
+        if not os.path.exists(full_path):
+            return False
+        try:
+            with open(full_path, "rb") as f:
+                disk_data = f.read()
+        except OSError:
+            return True
+        try:
+            blob = repo.object_store[index_entry.sha]
+        except KeyError:
+            return True
+        norm_data = normalizer.checkin_normalize(disk_data, full_path)
+        return norm_data != blob.data
+
+    # 1) Update skip-worktree bits
+    for path_bytes, entry in list(index.items()):
+        path_str = path_bytes.decode("utf-8")
+        if path_str in included_paths:
+            entry.set_skip_worktree(False)
+        else:
+            entry.set_skip_worktree(True)
+        index[path_bytes] = entry
+    index.write()
+
+    # 2) Reflect changes in the working tree
+    for path_bytes, entry in list(index.items()):
+        full_path = os.path.join(repo.path, path_bytes.decode("utf-8"))
+
+        if entry.skip_worktree:
+            # Excluded => remove if safe
+            if os.path.exists(full_path):
+                if not force and local_modifications_exist(full_path, entry):
+                    raise CheckoutError(
+                        f"Local modifications in {full_path} would be overwritten "
+                        "by sparse checkout. Use force=True to override."
+                    )
+                try:
+                    os.remove(full_path)
+                except IsADirectoryError:
+                    pass
+                except FileNotFoundError:
+                    pass
+        else:
+            # Included => materialize if missing
+            if not os.path.exists(full_path):
+                try:
+                    blob = repo.object_store[entry.sha]
+                except KeyError:
+                    raise Error(f"Blob {entry.sha} not found for {path_bytes}.")
+                ensure_dir_exists(os.path.dirname(full_path))
+                with open(full_path, "wb") as f:
+                    f.write(blob.data)
+
+
+class ParsedPattern:
+    """Represents one gitignore-style pattern (for sparse checkout).
+
+    We store:
+      - is_neg: whether the pattern is negative (leading '!')
+      - is_dir_only: whether it ends with '/'
+      - anchored: whether it starts with '/'
+      - regex: a compiled regex object to match against a path
+    """
+
+    __slots__ = (
+        "anchored",
+        "is_dir_only",
+        "is_neg",
+        "regex",
+    )
+
+    def __init__(self, is_neg, is_dir_only, anchored, regex):
+        self.is_neg = is_neg
+        self.is_dir_only = is_dir_only
+        self.anchored = anchored
+        self.regex = regex
+
+
+def fnmatch_to_regex(pat):
```

Comment:
Isn't this the same as ignore patterns? Can we reuse that code?

Reply:
So the gitignore style patterns (a.k.a. "full") are the simple ones, we want to also support "cone mode" patterns, so no, you're right though we don't need it all, some of this is getting deleted now...

Reply:
git docs on full vs. cone mode patterns are here:
- https://git-scm.com/docs/git-sparse-checkout#_internalsfull_pattern_set
- https://git-scm.com/docs/git-sparse-checkout#_internalscone_pattern_set